### PR TITLE
CLJS support

### PR DESCRIPTION
**Pull request by aengelberg**
- Most source and all test files now use `.cljc` format.
- Wrote [cljsee](https://github.com/aengelberg/cljsee) plugin and added it to instaparse, to keep the library compatible with Clojure 1.6 and lower.
- Re-implementations of key pieces of the Instaparse engine in ClojureScript. Credit: @mullr
- A ClojureScript test runner that runs on Node.js. Credit: @mullr @lbradstreet
- 2+ years of patching and iterative development. Credit: mostly @lbradstreet

### CODE REVIEW

File: project.clj [hunk 1]
```diff
@@ -1,19 +1,52 @@
...
+  :aliases {"test-all" ["do"
+                        "cljsee" "once,"
+                        "with-profile" "+1.5:+1.6:+1.7:+1.8:+1.9" "test"]
+            "test-cljs" ["cljsbuild" "test" "unit-tests"]
+            "test-cljs-all" ["with-profile" "+1.7:+1.8:+1.9" "do" "clean," "test-cljs"]}
+  :test-paths ["test/" "target/generated/test/clj"]
+  :source-paths ["src/" "target/generated/src/clj"]
+  :cljsee {:builds [{:source-paths ["src/"]
+                     :output-path "target/generated/src/clj"
+                     :rules :clj}
```
Comment (danielcompton) [inline project.clj]: Could that be added as a lein hook?

Comment (arrdem) [inline project.clj]: Yeah you can add it to the `lein release` task IIRC.

File: src/instaparse/core.cljc [hunk 2]
```diff
@@ -115,58 +120,73 @@
...
+     (invoke [parser text key1 val1 key2 val2 key3 val3] (parse parser text key1 val1 key2 val2 key3 val3))
+     (applyTo [parser args] (apply parse parser args))]
 
+    :cljs
+    [IFn
+     (-invoke [parser text] (parse parser text))
+     (-invoke [parser text key1 val1] (parse parser text key1 val1))
+     (-invoke [parser text key1 val1 key2 val2] (parse parser text key1 val1 key2 val2))
+     (-invoke [parser text key1 val1 key2 val2 key3 val3] (parse parser text key1 val1 key2 val2 key3 val3))
+     #_(-applyTo [parser args] (apply parse parser args))]))
```
Comment (Engelberg) [inline src/instaparse/core.cljc]: Need to add more -invoke lines in clojurescript since there is no applyTo.

File: src/instaparse/failure.cljc [hunk 3]
```diff
@@ -12,11 +12,25 @@
...
-    (catch Exception e "")))
+#?(:clj
+   (defn get-line
+     "Returns nth line of text, 1-based"
+     [n text]
+     (try (nth (line-seq (BufferedReader. (StringReader. (str text)))) (dec n))
+          (catch Exception e "")))
+   :cljs
+   (do (defn newline-chars? [c] 
+         (boolean (#{\newline \return} c)))
```
Comment (Engelberg) [inline src/instaparse/failure.cljc]: Remove \return

File: src/instaparse/failure.cljc [hunk 4]
```diff
@@ -42,16 +56,17 @@
...
     (print (print/regexp->str r))
     :else
     (pr r)))
 
 (defn pprint-failure
   "Takes an augmented failure object and prints the error message"
   [{:keys [line column text reason]}]
-  (printf "Parse error at line %d, column %d:\n"
-          line column)
+  (println "Parse error at line" line ", column" column ":\n")
```
Comment (Engelberg) [inline src/instaparse/failure.cljc]: Remove \n or change println to print.

File: README.md [hunk 5]
```diff
@@ -18,7 +18,7 @@ Instaparse aims to be the simplest way to build parsers in Clojure.
 
 ## Quickstart
 
-Instaparse requires Clojure v1.5.1 or later.  (It may work with earlier versions, but its speed relies extensively on features new to 1.5.)  A [Clojurescript port](https://github.com/lbradstreet/instaparse-cljs) is also available.
+Instaparse requires Clojure v1.5.1 or later, or ClojureScript v0.0-3196 or later.
```
Comment (Engelberg) [inline README.md]: Add any info to the README about things that aren't available in Clojurescript.

**Comment by lbradstreet** [comment]
@angelberg congratulations, and thanks for bringing this home!

**Pull request merged**